tinygrad/ops.py

```diff
@@ -332,16 +362,52 @@ def r(self, op:Ops, axis:Tuple[int, ...]):
   def assign(self, x:UOp): return UOp(Ops.ASSIGN, self.dtype, (self,x), None if self.st is None or self.st.contiguous else self.st)
-  def contiguous(self): return UOp(Ops.CONTIGUOUS, self.dtype, (self,))
+  def contiguous(self, allow_buffer_view=True):
+    if not unwrap(self.st).contiguous or self.size != self.base.size or self.is_unrealized_const():
+      if allow_buffer_view and self.can_view(): return self.metaop(Ops.BUFFER_VIEW, self.shape, self.dtype, self.device, None, (self,))
+      return self.alu(Ops.CONTIGUOUS)
+    forced_realize.add(self.base)
+    return self
 
   # *** from LazyBuffer ***
 
   @staticmethod
   def const_with_shape(dtype:DType, val:ConstLike, shape:Tuple[sint,...]) -> UOp:
     from tinygrad.shape.shapetracker import ShapeTracker
     return UOp(Ops.VALID, dtypes.bool, (ShapeTracker.from_shape(()).reshape((1,)*len(shape)).expand(shape).to_uop(),)).where(UOp.const(dtype, val), 0)
-  def is_unrealized_const(self): return (s:=self.base).op is Ops.VIEW and len(s.src) == 2 and s.src[1].op is Ops.CONST
+  @staticmethod
+  def metaop(op:Ops, shape:Tuple[sint, ...], dtype:DType, device:str, arg=None, src:Tuple[UOp, ...]=()) -> UOp:
+    from tinygrad.shape.shapetracker import ShapeTracker
+    # NOTE: we embed device on CONST with a fake BUFFER uop
+    if op is Ops.CONST:
+      return UOp(Ops.VIEW, dtype, (UOp.new_buffer(device, 1, dtype), arg if isinstance(arg, UOp) else UOp.const(dtype, unwrap(arg))),
+                 ShapeTracker.from_shape(())).reshape((1,)*len(shape)).expand(shape)
+    # otherwise it's a contiguous st
+    return UOp(Ops.VIEW, dtype, (UOp.new_buffer(device, (st:=ShapeTracker.from_shape(shape)).size, dtype), UOp(op, dtype, src, arg)), st)
+  def copy_to_device(self, device:str, force=False, clone:bool=False) -> UOp:
+    # no COPY
+    if self.device == device and not clone: return self
+    # if it's a shrink, do the shrink before the copy with CONTIGUOUS
+    if prod(self.shape) < prod(self.base.shape): return self.contiguous().copy_to_device(device)
+    # copy the base and apply the shapetracker on the new device
+    if not unwrap((src:=self.base).st).contiguous: raise RuntimeError(f"can only copy contiguous {self}")
+    return UOp.metaop(Ops.COPY, src.shape, src.dtype, device, src.nbytes, (src,)).view(unwrap(self.st))
+  def clone(self) -> UOp: return self.copy_to_device(self.device, clone=True)
+  def is_unrealized_const(self): return (s:=self.base).op is Ops.VIEW and len(s.src) == 2 and s.realized is None and s.src[1].op is Ops.CONST
   def is_unrealized_unmasked_const(self): return self.is_unrealized_const() and all(v.mask is None for v in unwrap(self.st).views)
+  def can_view(self):
+    return (self.st is not None and self._device is not None and self.st.consecutive and not self.is_unrealized_const() and
+            not isinstance(self.dtype, ImageDType) and self.device.split(":")[0] in view_supported_devices)
+  @property
+  def srcs(self): return self.src
+  @srcs.deleter
```

Comment:
Looks like this isn't used

Reply:
tensor uses it. for EMPTY fake realize hack.